File: src/Copilot/Companion/CompanionCipher.cs
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Modes;
using Org.BouncyCastle.Crypto.Parameters;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Web.Script.Serialization;

namespace Msfs2024Ai.Copilot.Companion;

internal sealed class CompanionCipher
{
    public const int WireVersion = 1;
    private const int NonceSize = 12;
    private const int TagBits = 128;
    private static readonly TimeSpan MaximumClockDifference =
        TimeSpan.FromMinutes(10);

    private readonly string _sessionId;
    private readonly byte[] _key;
    private readonly string _sender;
    private readonly string _expectedSender;
    private readonly HashSet<string> _receivedMessageIds =
        new(StringComparer.Ordinal);
    private readonly Queue<string> _receivedMessageOrder = new();

    public CompanionCipher(
        string sessionId,
        byte[] pairingSecret,
        string sender)
    {
        if (pairingSecret.Length != 32)
        {
            throw new ArgumentException(
                "The pairing secret must contain exactly 32 bytes.",
                nameof(pairingSecret));
        }
        if (sender != "desktop" && sender != "tablet")
        {
            throw new ArgumentException(
                "The companion cipher sender must be desktop or tablet.",
                nameof(sender));
        }

        _sessionId = sessionId;
        _key = Derive(pairingSecret, "vfo-e2e-key-v1:" + sessionId);
        _sender = sender;
        _expectedSender = sender == "desktop" ? "tablet" : "desktop";
    }

    public string Seal(string plaintext) =>
        Seal(plaintext, Guid.NewGuid().ToString("N"), DateTimeOffset.UtcNow, null);

    internal string Seal(
        string plaintext,
        string messageId,
        DateTimeOffset sentUtc,
        byte[]? suppliedNonce)
    {
        var nonce = suppliedNonce ?? RandomBytes(NonceSize);
        if (nonce.Length != NonceSize)
        {
            throw new ArgumentException("The nonce must contain 12 bytes.", nameof(suppliedNonce));
        }
        var sentUnixMillis = sentUtc.ToUnixTimeMilliseconds();
        var aad = BuildAdditionalData(
            _sender,
            messageId,
            sentUnixMillis);
        var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
        var cipher = new ChaCha20Poly1305();
        cipher.Init(
            true,
            new AeadParameters(
                new KeyParameter(_key),
                TagBits,
                nonce,
                aad));
        var ciphertext = new byte[cipher.GetOutputSize(plaintextBytes.Length)];
        var written = cipher.ProcessBytes(
            plaintextBytes,
            0,
            plaintextBytes.Length,
            ciphertext,
            0);
        written += cipher.DoFinal(ciphertext, written);
        if (written != ciphertext.Length)
        {
            Array.Resize(ref ciphertext, written);
        }

        return CompanionProtocol.Serialize(
            new Dictionary<string, object>
            {
                ["wireVersion"] = WireVersion,
                ["sender"] = _sender,
                ["messageId"] = messageId,
                ["sentUnixMillis"] = sentUnixMillis,
                ["nonce"] = Base64Url.Encode(nonce),
                ["ciphertext"] = Base64Url.Encode(ciphertext)
            });
    }

    public bool TryOpen(string envelope, out string plaintext, out string error)
    {
        plaintext = string.Empty;
        error = string.Empty;
        IDictionary<string, object>? values;
        try
        {
            values = new JavaScriptSerializer().DeserializeObject(envelope)
                as IDictionary<string, object>;
        }
        catch
        {
            error = "The encrypted companion envelope was not valid JSON.";
            return false;
        }

        if (values == null
            || !TryReadInt(values, "wireVersion", out var wireVersion)
            || wireVersion != WireVersion
            || !string.Equals(ReadString(values, "sender"), _expectedSender, StringComparison.Ordinal)
            || !TryReadLong(values, "sentUnixMillis", out var sentUnixMillis))
        {
            error = "The encrypted companion envelope header was invalid.";
            return false;
        }
        var messageId = ReadString(values, "messageId");
        if (string.IsNullOrWhiteSpace(messageId)
            || messageId == null
            || messageId.Length > 80
            || _receivedMessageIds.Contains(messageId))
        {
            error = "The encrypted companion message was missing, invalid, or already received.";
            return false;
        }

        DateTimeOffset sentUtc;
        try
        {
            sentUtc = DateTimeOffset.FromUnixTimeMilliseconds(sentUnixMillis);
        }
        catch (ArgumentOutOfRangeException)
        {
            error = "The encrypted companion message timestamp was invalid.";
            return false;
        }
        if ((DateTimeOffset.UtcNow - sentUtc).Duration() > MaximumClockDifference)
        {
            error = "The encrypted companion message was stale or the device clocks differ too much.";
            return false;
        }

        if (!Base64Url.TryDecode(ReadString(values, "nonce") ?? string.Empty, out var nonce)
            || nonce.Length != NonceSize
            || !Base64Url.TryDecode(
                ReadString(values, "ciphertext") ?? string.Empty,
                out var ciphertext)
            || ciphertext.Length < TagBits / 8)
        {
            error = "The encrypted companion message data was invalid.";
            return false;
        }

        try
        {
            var cipher = new ChaCha20Poly1305();
            cipher.Init(
                false,
                new AeadParameters(
                    new KeyParameter(_key),
                    TagBits,
                    nonce,
                    BuildAdditionalData(
                        _expectedSender,
                        messageId,
                        sentUnixMillis)));
            var output = new byte[cipher.GetOutputSize(ciphertext.Length)];
            var written = cipher.ProcessBytes(
                ciphertext,
                0,
                ciphertext.Length,
                output,
                0);
            written += cipher.DoFinal(output, written);
            plaintext = Encoding.UTF8.GetString(output, 0, written);
        }
        catch (InvalidCipherTextException)
        {
            error = "The encrypted companion message could not be authenticated.";
            return false;
        }

        RememberMessage(messageId);
        return true;
    }

    public static string DeriveRelayCredential(
        string sessionId,
        byte[] pairingSecret) =>
        Base64Url.Encode(Derive(
            pairingSecret,
            "vfo-relay-auth-v1:" + sessionId));

    private byte[] BuildAdditionalData(
        string sender,
        string messageId,
        long sentUnixMillis) =>
        Encoding.UTF8.GetBytes(
            $"{WireVersion}|{_sessionId}|{sender}|{messageId}|{sentUnixMillis}");

    private void RememberMessage(string messageId)
    {
        _receivedMessageIds.Add(messageId);
        _receivedMessageOrder.Enqueue(messageId);
        while (_receivedMessageOrder.Count > 512)
        {
            _receivedMessageIds.Remove(_receivedMessageOrder.Dequeue());
        }
    }

    private static byte[] Derive(byte[] secret, string purpose)
    {
        using var hmac = new HMACSHA256(secret);
        return hmac.ComputeHash(Encoding.UTF8.GetBytes(purpose));
    }

    private static byte[] RandomBytes(int count)
    {
        var bytes = new byte[count];
        using var random = RandomNumberGenerator.Create();
        random.GetBytes(bytes);
        return bytes;
    }

    private static string? ReadString(
        IDictionary<string, object> values,
        string key) =>
        values.TryGetValue(key, out var value) ? value as string : null;

    private static bool TryReadInt(
        IDictionary<string, object> values,
        string key,
        out int result)
    {
        result = 0;
        try
        {
            if (!values.TryGetValue(key, out var value) || value == null)
            {
                return false;
            }
            result = Convert.ToInt32(value);
            return true;
        }
        catch
        {
            return false;
        }
    }

    private static bool TryReadLong(
        IDictionary<string, object> values,
        string key,
        out long result)
    {
        result = 0;
        try
        {
            if (!values.TryGetValue(key, out var value) || value == null)
            {
                return false;
            }
            result = Convert.ToInt64(value);
            return true;
        }
        catch
        {
            return false;
        }
    }
}

File: src/AndroidCompanion/app/src/main/java/com/noscapect/vfo/companion/security/CompanionCipher.kt
package com.noscapect.vfo.companion.security

import java.nio.charset.StandardCharsets
import java.security.GeneralSecurityException
import java.security.SecureRandom
import java.util.UUID
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class CompanionCipher(
    private val sessionId: String,
    pairingSecret: ByteArray,
    private val sender: String,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val key = derive(pairingSecret, "vfo-e2e-key-v1:$sessionId")
    private val expectedSender = if (sender == "desktop") "tablet" else "desktop"
    private val receivedIds = LinkedHashSet<String>()

    init {
        require(pairingSecret.size == 32) { "Pairing secret must contain 32 bytes." }
        require(sender == "desktop" || sender == "tablet") {
            "Companion sender must be desktop or tablet."
        }
    }

    fun seal(plaintext: String): String = seal(
        plaintext = plaintext,
        messageId = UUID.randomUUID().toString().replace("-", ""),
        sentUnixMillis = System.currentTimeMillis(),
        nonce = ByteArray(NONCE_SIZE).also(SecureRandom()::nextBytes),
    )

    internal fun seal(
        plaintext: String,
        messageId: String,
        sentUnixMillis: Long,
        nonce: ByteArray,
    ): String {
        require(nonce.size == NONCE_SIZE) { "Nonce must contain 12 bytes." }
        val cipher = Cipher.getInstance("ChaCha20-Poly1305")
        cipher.init(
            Cipher.ENCRYPT_MODE,
            SecretKeySpec(key, "ChaCha20"),
            IvParameterSpec(nonce),
        )
        cipher.updateAAD(additionalData(sender, messageId, sentUnixMillis))
        val ciphertext = cipher.doFinal(plaintext.toByteArray(StandardCharsets.UTF_8))
        return json.encodeToString(
            EncryptedEnvelope(
                sender = sender,
                messageId = messageId,
                sentUnixMillis = sentUnixMillis,
                nonce = Base64Url.encode(nonce),
                ciphertext = Base64Url.encode(ciphertext),
            )
        )
    }

    @Synchronized
    fun open(envelopeJson: String): Result<String> = runCatching {
        val envelope = json.decodeFromString<EncryptedEnvelope>(envelopeJson)
        require(envelope.wireVersion == WIRE_VERSION) { "Unsupported encrypted wire version." }
        require(envelope.sender == expectedSender) { "Unexpected encrypted message sender." }
        require(envelope.messageId.isNotBlank() && envelope.messageId.length <= 80) {
            "Invalid encrypted message ID."
        }
        require(envelope.messageId !in receivedIds) { "Encrypted message was already received." }
        require(kotlin.math.abs(System.currentTimeMillis() - envelope.sentUnixMillis) <= MAX_CLOCK_DIFFERENCE_MS) {
            "Encrypted message is stale or the device clocks differ too much."
        }

        val nonce = Base64Url.decode(envelope.nonce)
        val ciphertext = Base64Url.decode(envelope.ciphertext)
        require(nonce.size == NONCE_SIZE && ciphertext.size >= TAG_SIZE) {
            "Invalid encrypted message data."
        }
        val cipher = Cipher.getInstance("ChaCha20-Poly1305")
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(key, "ChaCha20"),
            IvParameterSpec(nonce),
        )
        cipher.updateAAD(
            additionalData(
                expectedSender,
                envelope.messageId,
                envelope.sentUnixMillis,
            )
        )
        val plaintext = cipher.doFinal(ciphertext).toString(StandardCharsets.UTF_8)
        receivedIds += envelope.messageId
        while (receivedIds.size > MAX_RECEIVED_IDS) {
            receivedIds.remove(receivedIds.first())
        }
        plaintext
    }

    private fun additionalData(
        messageSender: String,
        messageId: String,
        sentUnixMillis: Long,
    ): ByteArray =
        "$WIRE_VERSION|$sessionId|$messageSender|$messageId|$sentUnixMillis"
            .toByteArray(StandardCharsets.UTF_8)

    companion object {
        const val WIRE_VERSION = 1
        private const val NONCE_SIZE = 12
        private const val TAG_SIZE = 16
        private const val MAX_CLOCK_DIFFERENCE_MS = 10 * 60 * 1000L
        private const val MAX_RECEIVED_IDS = 512

        fun deriveRelayCredential(sessionId: String, secret: ByteArray): String {
            require(secret.size == 32) { "Pairing secret must contain 32 bytes." }
            return Base64Url.encode(derive(secret, "vfo-relay-auth-v1:$sessionId"))
        }

        private fun derive(secret: ByteArray, purpose: String): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(secret, "HmacSHA256"))
            return mac.doFinal(purpose.toByteArray(StandardCharsets.UTF_8))
        }
    }
}

@Serializable
private data class EncryptedEnvelope(
    val wireVersion: Int = CompanionCipher.WIRE_VERSION,
    val sender: String,
    val messageId: String,
    val sentUnixMillis: Long,
    val nonce: String,
    val ciphertext: String,
)

object Base64Url {
    fun encode(value: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(value)

    fun decode(value: String): ByteArray {
        require(value.isNotBlank() && value.all { it.isLetterOrDigit() || it == '-' || it == '_' }) {
            "Invalid base64url value."
        }
        return try {
            Base64.getUrlDecoder().decode(value)
        } catch (exception: IllegalArgumentException) {
            throw GeneralSecurityException("Invalid base64url value.", exception)
        }
    }
}
